sympy/assumptions/refine.py

```diff
@@ -240,6 +240,73 @@ def refine_Relational(expr, assumptions):
     return ask(Q.is_true(expr), assumptions)
 
 
+def refine_re(expr, assumptions):
+    """
+    Handler for real part.
+
+    >>> from sympy.assumptions.refine import refine_re
+    >>> from sympy import Q, re
+    >>> from sympy.abc import x
+    >>> refine_re(re(x), Q.real(x))
+    x
+    >>> refine_re(re(x), Q.imaginary(x))
+    0
+    """
+    from sympy.functions.elementary.complexes import re, im
+    arg = expr.args[0]
+    if arg.is_Atom:
+        if ask(Q.real(arg), assumptions):
+            return arg
+        if ask(Q.imaginary(arg), assumptions):
+            return 0
+    if isinstance(arg, Mul):
```

Comment:
After the two cases above I wonder if the next step should be something like `.as_real_imag()`. Here you special case Mul but you would end up with a lot of cases that duplicate code from other places if you included everything...

Reply:
I am sorry, I did not understand what you meant. Can you elaborate a bit more?

Reply:
I guess I'm not really sure what is reasonable to implement in refine. Maybe someone else can comment on this PR...

Reply:
Yeah. I guess it's better to leave this PR be unmerged. There seem be many cases where there is no general way of simplifying real, complex part.

Reply:
This PR seems good. I'm just unclear about the development goals for refine myself.

What I'm not sure about is this: there already is a lot of code for handling Mul, Pow, exp, erf, ... and for doing things like working out if they getting imaginary parts or checking if something is real or nonzero etc. and I don't know to what extent refine should reuse those things without reimplementing everything.

Reply:
Makes sense, no point rewriting all those parts.